src/AutoFakerBinder.cs

```diff
@@ -344,6 +344,10 @@ private static void PopulateCollection(object value, object parent, AutoMember m
             return;
 
         object? instance = member.Getter(parent);
+
+        if (instance is IList { IsReadOnly: true } or IDictionary { IsReadOnly: true })
```

Comment:
I think you can simplify and speed things up by checking member.IsReadOnly before the instance is even retrieved via the getter?

Reply:
With the latest version of reflection cache, you'll need to also do this readonly checking in the PopulateDictionary() method too

Reply:
> I think you can simplify and speed things up by checking member.IsReadOnly before the instance is even retrieved via the getter?

Unfortunately, this is not possible when the member type is `IList` or `IDictionary` because the implementation decides whether it wants the instance to be read-only when returned. There is nothing preventing the state of the object, or some other condition from being used to determine whether the returned instance should be read-only.

```csharp
public class Foo
{
	private readonly List<string> _bars = new();

	public IList<string> Bars => DateTime.UtcNow.Ticks % 2 == 0 ? _bars : _bars.AsReadOnly();
}
```

Not a friendly way to code things but we don't control all the types we consume.

Reply:
> With the latest version of reflection cache, you'll need to also do this readonly checking in the PopulateDictionary() method too

I wondered why that code path wasn't executed. I've refactored the code with two separate checks.

Reply:
@alexmg I think I understand what you're saying, but `AutoMember.IsReadOnly` isn't just checking if the field is `init-only`, it's checking for if the property `can write`, which I believe in this scenario, it still would be able to determine whether that's true or not, right?

I'm happy to be wrong, just trying to understand

Reply:
I think the `AutoMember.IsReadOnly` property name is fine. 😄 

I have updated the unit tests to better demonstrate the issue by including an example of what should happen when the collection or dictionary returned from the property is not wrapped using the `AsReadOnly` methods at runtime.

There are now two properties on the test types that look identical except one wraps the returned instance using the `AsReadOnly` method and the other does not. Both properties on the test types are not `CanWrite` but do return a collection or dictionary resulting in the `PopulateCollection` or `PopulateDictionary` method being called. The returned instance **could** have members added to it even though the property itself is not writable.

The interface types `ICollection` and `IDictionary` are not considered read-only in terms of reflection like `IReadOnlyCollection<T>` or `IReadOnlyDictionary<TKey,TValue>`. When a mutable collection is wrapped in a `ReadOnlyCollection<T>` or `ReadOnlyDictionary<TKey, TValue>` at runtime the `IsReadOnly` will return `true` and calls to the `Add` method will throw an exception.

Reflection cannot be used to check whether `ICollection` or `IDictionary` will throw an exception when the `Add` method is called at runtime. You need to explicitly check the runtime value of the `IsReadOnly` property to be sure it's safe to call the `Add` method.